### PR TITLE
fix: update post process logic for claude instant

### PR DESCRIPTION
**Pull request by abeatrix**
fix: use full infill block and adjust prefix

- fix prompt structure order, context should be shared after the role assigning
- empty prefix if it contains the infill block for issue where first line is duplicated 
- Simplify prompt messages to focus on providing clear context
- Add CLOSING tag for Assistant's infill block to preserve ending whitespace 
- This also fixed issues where multi-line autocomplete gets cut off.
- remove tags for context code as that might cause confusion for claude

## Test plan

Fix: show multi-line completion after `{`:

image

Fix: show multi-line completions the line after `{`

image

### CODE REVIEW

**Review by valerybugakov — COMMENTED** [review]
Could we separate the bug fix from the prompt changes and put the prompt changes behind the feature flag so that we can test how they affect CAR in production before making it a default? 

The prompt updates look logical, and I expect them to perform well, but with LLMs, we never know how newly added tokens will affect the output. We have quite a few changes here, from the sentence structure updates to changing the code block tags. Splitting this PR in two would make it easier to reason about CAR changes later:

1. Minimal prompt change required to fix a bug + logic updates.
2. Prompt updates behind the feature flag.

**Comment by abeatrix** [comment]
@valerybugakov that's fair, I can do that next!

Have you tried the change in this PR yet? Is it working for you?

**Comment by philipp-spiess** [comment]
TBH I’m happy to move fast with this given how big our regression is right now. My only thinking is that:

- we should run our completion test suite to ensure we don't add known issues
- we delete old code that is no longer needed (like the head/tail splitting of the prefix) to avoid introducing code dept

once you're happy with the results we should push out a patch release.

**Review by philipp-spiess — APPROVED** [review]
preemptive stamp but lets make sure the tests are good

**Comment by abeatrix** [comment]
@philipp-spiess @valerybugakov I tried to make minimal prompt change as possible but the one we had just doesn't work anymore because of the lines and whitespaces removed on claude side. I added <_cursor_> to the end of the infill block to make sure the spaces are not removed, and it seems to work fine on my side, but i wasn't able to get the eval tests to work to confirm if there is other regression, but definately better than what is in prod right now. If this is good for now, please feel free to merge and make a patch release on monday to make sure our users are not getting affected, and I can do any required follow up works

**Comment by philipp-spiess** [comment]
@abeatrix I pushed a fix for the local completions test and there seem to be some issues with the newly used XML leaking into the completion:

Screenshot 2023-10-23 at 11 37 46

Also the test case with a comment followed by a `\n` isn't working well either :/ 

Screenshot 2023-10-23 at 11 38 47

**Comment by abeatrix** [comment]
@valerybugakov After taking a look at the issue with @philipp-spiess ,  it looks like the root cause of the issue was the new `parseAndTruncateCompletion` that doesn't work well with the `this.postProcess()` for anthropic. I tried added some workaround in my lastest commits and that seems to work (see the updated infill test result). Can you take a look and see if this change makes sense to you? Since this is a P1 issue, do you mind taking over of this PR while I'm away to make sure we can have this issue resolved asap?

File: vscode/src/completions/providers/anthropic.ts [hunk 1]
```diff
@@ -67,7 +67,7 @@ export class AnthropicProvider extends Provider {
         const { head, tail, overlap } = getHeadAndTail(this.options.docContext.prefix)
 
         // Infill block represents the code we want the model to complete
-        const infillBlock = tail.trimmed
+        const infillBlock = tail.trimmed.endsWith('{\n') ? tail.trimmed.trimEnd() : tail.trimmed
```
Comment (abeatrix) [inline vscode/src/completions/providers/anthropic.ts:70]: this allows claude to response with multi-line completion on new line after the `{` bracket

File: vscode/src/completions/providers/anthropic.ts [hunk 2]
```diff
@@ -216,6 +216,10 @@ export class AnthropicProvider extends Provider {
             // leading `\n` followed by whitespace that Claude might add.
             completion = completion.replace(/^\s*\n\s*/, '')
         } else {
+            // prevent normalizeStartLine from removing the starting new line
+            if (completion.startsWith('\n')) {
+                completion = '\n' + completion.trimStart()
```
Comment (abeatrix) [inline vscode/src/completions/providers/anthropic.ts:221]: `normalizeStartLine` in `parseAndTruncateCompletion` would remove the `\n` if it follows by indentation when prefix didn't end with closing bracket 🤔  This solves the aforementioned issue

**Comment by valerybugakov** [comment]
@abeatrix, thank you for sharing the context! I'm debugging it locally and will merge this if I can confirm your findings 👍

File: vscode/src/completions/providers/anthropic.ts (hunk 2 above)
Comment (valerybugakov) [reply vscode/src/completions/providers/anthropic.ts]: Updated the logic to `normalizeStartLine` only for multiline completions like before my changes in [this PR](https://github.com/sourcegraph/cody/pull/1404/files). It fixes the issue without adding the additional logic. Thank you for narrowing down this regression!

**Review by valerybugakov — APPROVED** [review]
After my change, I verified that the cases mentioned here still work locally as expected. Regenerated completions using the `generate:completions` to ensure we do not have any leaking XML tags.

**Comment by valerybugakov** [comment]
The PR is blocked by eslint errors on main. Fixing it here: https://github.com/sourcegraph/cody/pull/1471

**Pull request merged**